### ISSUE TITLE
🐞 Dagger sends cached context even though file is missing locally

### ISSUE THREAD

**Issue by marcosnils** [dagger/dagger#4757]
### What is the issue?

I was running a pipeline which mounts a binary from my host into Dagger but after removing the binary locally, my pipeline still executed with a cached version of it

### Log output

N/A

### Steps to reproduce

1 - Build a static binary in your CWD called main
2 - Run the following pipeline once: 

```go 
package main

import (
	"context"
	"fmt"
	"os"
	"time"

	"dagger.io/dagger"
)

func main() {
	ctx := context.Background()
	c, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))
	if err != nil {
		panic(err)
	}

	bin := c.Host().Workdir(dagger.HostWorkdirOpts{
		Include: []string{"main"},
	}).File("main")

	out, err := c.Container().
		WithEnvVariable("CACHE", time.Now().String()).
		WithMountedFile("/main", bin).
		WithExec([]string{"/main"}).ExitCode(ctx)
	if err != nil {
		panic(err)
	}

	fmt.Println(out)
}
```
3 - Remove the `main` binary after the pipeline succeeds. 
4 - Running the pipeline again will pass when I expected it to fail.

This is what I see in the output: 
```
#1 host.directory /home/marcos/Projects/testbug
#1 transferring /home/marcos/Projects/testbug: 10.10MB 0.1s done
#1 DONE 0.1s

#2 importing cache manifest from dagger:10686922502337221602
#2 DONE 0.0s

```

FWIW this error also happens with current [tip](55504a723d28d8c2a578ac7936a8ace827bea84f) verison. Another interesting fact is that even if I bust the engine cache with `buildctl prune`, this keeps happening.

### SDK version

Go v0.5.0

### OS version

Linux xps 5.19.0-35-generic #36-Ubuntu SMP PREEMPT_DYNAMIC Fri Feb 3 18:36:56 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

**Comment by mircubed** [comment]
Confirmed with @sipsma that this isn't the correct behavior. Assigned to @marcosnils for further investigation.

**Comment by sipsma** [comment]
The first question that's worth looking into is whether this is A) a result of the dir synchronization not copying the delete of the local file to the remote or B) a result of the delete in the remote snapshot not invalidating the cache of the subsequent steps.

This is not super straightforward to debug. My immediate thought would be to
1. start with a completely clean engine (i.e. `docker volume rm dagger-engine.dev`)
2. run the repro once
3. look at buildkit's internal snapshot dirs and find the ones with the synced local content
   * `docker exec -it dagger-engine.dev ls -la /var/lib/dagger/runc-overlayfs/snapshots/snapshots` will list all the snapshots
   * I'd expect there to be two snapshots: 1 is the actual sync'd dir and one is the copy we make as described [here](https://github.com/dagger/dagger/blob/d3c6dd9987b7896ecf2de7135196154415391c79/core/host.go#L90-L98)
4. run the repro again
5. look at buildkit's internal snapshot dirs again
   * The expected behavior should be that the underlying sync'd local dir snapshot should no longer have the file which was deleted and there should be a copy of that which is the same. However, given the bug here, it seems like that should probably not end up being the case.

**Comment by TomChv** [comment]
Hey, I'll continue the investigation and let you know guys

**Comment by TomChv** [comment]
@sipsma @marcosnils It's strange because I cannot reproduce the bug :(

```shell
# Run repro with main
go run ./test_4757_repro.go                          
#1 host.directory /Users/tomchauveau/Documents/DAGGER/dagger
#1 transferring /Users/tomchauveau/Documents/DAGGER/dagger: 27B done
#1 DONE 0.0s

#2 
#2 0.151 hello world!
#2 DONE 0.2s

#3 importing cache manifest from dagger:10686922502337221602
0

# RM binary
rm ./main 

# Rerun repro
go run ./test_4757_repro.go
#1 importing cache manifest from dagger:10686922502337221602
#1 DONE 0.0s

#2 host.directory /Users/tomchauveau/Documents/DAGGER/dagger
#2 transferring /Users/tomchauveau/Documents/DAGGER/dagger: 2B done
#2 DONE 0.0s

#3 
#0 0.087 runc run failed: unable to start container process: error during container init: error mounting "/var/lib/dagger/runc-overlayfs/cachemounts/buildkit1253059466/main" to rootfs at "/main": stat /var/lib/dagger/runc-overlayfs/cachemounts/buildkit1253059466/main: no such file or directory
#3 ERROR: process "./main" did not complete successfully: exit code: 1
panic: input:1: container.withEnvVariable.withMountedFile.withExec.exitCode process "./main" did not complete successfully: exit code: 1

```

**Comment by marcosnils** [comment]
@TomChv let me try again. Maybe it was a buildkit thing that was fixed in the latest releases?

**Comment by marcosnils** [comment]
@TomChv I effectively can't repro. Even going back to the version where I initially saw the bug. Fun fact is that I can't recall if @jlongtine or @vito also saw this, just to make sure I'm not crazy :loco:

In any case, I can't repro anymore so I'll close this for the moment :pray:
